teps/0130-pipeline-level-service.md

```diff
@@ -0,0 +1,759 @@
+  tasks:
+  - name: git-clone
+    ...
+  - name: docker-build
+    params:
+    - name: image
+    - name: docker-host
+    workspaces:
+    - name: source
+    taskSpec:
+      workingDir: $(workspaces.source.path)
+      steps:
+      - image: docker.io/library/docker
+        script: docker build -f ./Dockerfile -t $(params.image)
+        env:
+        - name: DOCKER_HOST
+          value: $(params.docker-host)
+      results:
+      - name: IMAGE_DIGEST
+    runAfter: ["git-clone"]
+  # This task uses the built image without first pushing it to a registry
+  - name: pytest
+    params:
+    - name: docker-host
+    - name: image
+    taskSpec:
+      workingDir: $(workspaces.source.path)
+      steps:
+      - image: docker.io/library/docker
+        script: docker exec $(params.image) python -m pytest
+        env:
+        - name: DOCKER_HOST
+          value: $(params.docker-host)
+    runAfter: ["docker-build"]
+  # This task pushes the image to the registry only if tests passed
+  - name: docker-push
+    when:
+    - input: "$(tasks.pytest.status)"
+      operator: in
+      values: ["Succeeded"]
+    params:
+    - name: image
+    - name: docker-host
+    taskSpec:
+      steps:
+      - image: docker.io/library/docker
+        script: docker push $(params.image)
+        env:
+        - name: DOCKER_HOST
+          value: $(params.docker-host)
+    runAfter: ["pytest"]
+  sidecars:
+  - name: docker-daemon
+    image: docker:dind
+    securityContext:
+      privileged: true
+    ports:
+    - 2376
+    startupProbe:
+      tcpSocket:
+        port: 2376
+    workspaces:
+    - name: docker-tls-certs   
+```
+
+The sidecar would likely be implemented as a deployment with one pod plus a service.
+If the sidecar has readiness or startup probes, the PipelineRun would wait for them to succeed before starting the Pipeline Tasks.
+We could use a [NetworkPolicy](https://kubernetes.io/docs/concepts/services-networking/network-policies/)
+to prevent other PipelineRuns from being able to use the sidecar.
+
+Pros:
+- Good separation of concerns: App developer creating build Pipeline doesn't have to worry about using the correct service accounts
+or kube credentials
+- Arguably better UX than using separate setup and teardown Tasks. Don't have to worry about resource leakage if teardown Task
+fails due to a misconfigured PipelineRun.
+
+Cons:
+- Sidecar configurations will likely be copy-pasted between Pipelines.
+- If docker TLS certs are used, they must now be bound in a persistent volume claim to share
+between multiple Tasks, rather than an emptyDir volume to share between multiple steps.
+This is a pretty heavyweight option for a relatively small amount of data.
+However, this is more an artifact of the limited options Pipelines currently has for sharing
+data between Tasks, which should be treated as a separate problem.
+
+#### Open questions
+
+- Should Pipeline-level sidecars terminate before or after Finally Tasks?
```

Comment:
Most likely after. They are created before any TaskRun, and they are deleted after any TaskRun.